Documentation for redesigned sidechains (#1516)
XRPLF/xrpl-dev-portal: content/concepts/interoperability/witness-server.md

---
html: witness-server.html
parent: interoperability.html
blurb: A witness server is a light-weight server that witnesses and signs transactions between the XRP Ledger and another chain. 
---
# Witness Server

> The _witness server_ is a light-weight server that is aware the locking and issuing chains in a bridging solution between blockchains. The witness server witnesses and signs transactions between [a locking chain and an issuing chain](cross-chain-bridges.md) when assets are moved to designation addresses, thus enabling cross-chain transactions. 

Reviewer: - Looks like the first sentence may be missing a word or two?
- Move the link with background reading to the first mention (also, links need to be to the HTML, not md, file)
- I think you mean "designated addresses" rather than "designation addresses" but I substituted "door" which I believe to be more specifically what you're referring to here.

```suggestion
The _witness server_ is a light-weight server that is part of a [cross-chain bridge](cross-chain-bridges.html) and is aware of both the locking and issuing chains in such a bridge. The witness server witnesses and signs transactions when assets are moved to door addresses, thus enabling cross-chain transactions. 
```

The witness server is an independent server that has similar responsibilities as that of a validator server on the XRP Ledger's peer-to-peer network and helps avoid double-spend and collusion. 

The witness server serves as a neutral witness for transactions between a locking chain and an issuing chain. 

> It listens to the door accounts on both sides of the bridge and signs attestations for cross-chain transfer transactions, essentially affirming that a transaction on the source account happened, so the value can be claimed on the destination account. They are essentially acting as an oracle, to “prove” that the value was locked/burned on the source account, which allows the recipient to then claim (via minting/ unlocking) the equivalent funds on the destination account.

Reviewer: This should continue line 12, I think.

If it were meant to be a new paragraph, you'd need a blank line in between. Having a random single newline in the middle of a paragraph does nothing in Markdown, and we don't generally hard-wrap text for this site.

The bridge between the locking chain and the issuing chain includes the following information in its configuration: 

* Witness servers that monitor transactions on the bridge. You can choose one or more witness servers. 
* Fee for witness servers for their service.
 
Anyone can run a witness server. However, the burden is on the participants of the issuing chain to evaluate the reliability of witness servers. 

> Note that an issuing chain may choose to configure a bridge with only one witness server initially and run the witness server itself. This strategy may be helpful in the initial period when the issuing chain is yet to establish itself in the marketplace.

Reviewer: How does "the chain...itself" run a server? Feels like there's a person/organization being elided there.

## Witness Server Configuration

The witness server takes a JSON configuration file with the `--config` file.

```json
{
  "mainchain_endpoint": {
    "ip": "127.0.0.1",

>     "port": 6005

Reviewer: I'm guessing these are JSON-RPC, WS, or peer protocol endpoints for rippled servers connected to each chain? Should probably clarify what the values in this config file mean.

---
html: witness-server.html
parent: xrpl-interoperability.html
blurb: A witness server is a light-weight server that witnesses and signs transactions between the XRP Ledger and another chain. 
---
# Witness Server

> The _witness server_ is a light-weight server that is aware the locking and issuing chains in a bridging solution between blockchains. The witness server witnesses and signs transactions between [a locking chain and an issuing chain](cross-chain-bridges.md) when assets are moved to designation addresses, thus enabling cross-chain transactions. 

Reviewer 2: ```suggestion
The _witness server_ is a light-weight server that is aware of the locking and issuing chains in a bridging solution between blockchains. The witness server witnesses and signs transactions between [a locking chain and an issuing chain](cross-chain-bridges.md) when assets are moved to designated addresses, thus enabling cross-chain transactions. 
```

The bridge between the locking chain and the issuing chain includes the following information in its configuration: 

* Witness servers that monitor transactions on the bridge. You can choose one or more witness servers. 
* Fee for witness servers for their service.
 
Anyone can run a witness server. However, the burden is on the participants of the issuing chain to evaluate the reliability of witness servers. 

Note that an issuing chain may choose to configure a bridge with only one witness server initially and run the witness server itself. This strategy may be helpful in the initial period when the issuing chain is yet to establish itself in the marketplace.

## Witness Server Configuration

The witness server takes a JSON configuration file, specified using the `--conf` command-line argument.

The configuration file contains the following information:

> * Websocket endpoints for the locking chain and the issuing chain.

Reviewer 2: ```suggestion
* Websocket endpoints for the locking chain and the issuing chain (`LockingChainEndpoint` and `IssuingChainEndpoint`).
```
ditto for the other ones

> * Port that clients can use to connect to this server ("RPCENdpoint").

Reviewer 2: ```suggestion
* Port that clients can use to connect to this server (`RPCEndpoint`).
```

* Directory to store the SQL database
* Secret key used to sign attestations

> * Information about door accounts on the locking and issuing chains.

Reviewer 2: These aren't properly bulleted in the preview

Here is an example configuration file for a witness server:

> ```json

Reviewer 2: This config file shape has changed a lot since this version.
Here's an example jinja for one: https://github.com/XRPLF/sidechain-cli/blob/main/sidechain_cli/server/config/templates/witness.jinja
